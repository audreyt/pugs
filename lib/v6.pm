package v6;

use strict;
use Inline; # this is going away soon
use Filter::Simple;

$v6::VERSION = '0.02';

FILTER {
    my $marker = 'XXXXXXXX';
    $marker++ while /^$marker$/m;
    $_ = <<END;
use Inline Pugs => <<'$marker';
$_
$marker
END
};

1;

__DATA__

=head1 NAME 

v6 - Pragma for using Perl 6 inside Perl 5

=head1 SYNOPSIS

    #!/usr/bin/perl
    use v6-pugs; # Here is some Perl 6 code...
    sub postfix:<!> { [*] 1..$_ }
    sub sum_factorial { [+] 0..$_! }

    no v6;       # Here is some Perl 5 code...
    print sum_factorial(3); # 21

=head1 DESCRIPTION

The B<v6> pragma lets you put Perl 6 code in Perl 5 source code
after a line of:

    use v6-pugs;

To switch back to Perl 5:

    no v6;

=head1 NOTES

Whilst resembling C<use VERSION> syntax, Perl 5 actually (mis)interprets
C<use v6-pugs> as a request to use a module named C<v6.pm>, passing in C<'-pugs'>
as an argument list to C<import> after C<v6.pm> is C<require>d.  The lack of
whitespace between C<v6> and C<-pugs> is neccessary in this instance,
due to the bizarre interaction between the hybrid C<use VERSION> and
C<use Module VERSION> syntax.

Larry gave the following explanation of why it works on C<#perl6>:

    ...the interaction of v-syntax with 'use' versions is interesting, to say
    the least, as witnessed by 'v6-pugs'.

    In theory, Perl 5 oughta just blow up the moment it sees a 'use v6', saying 
    "I'm not version 6..." But there's chicanery in the lexer to rearrange 
    module versions, such that the parser sees the version first, and that's 
    getting fooled by 'v6-pugs' into thinking it's a module version when the 
    parser thinks otherwise.

=head1 SEE ALSO

L<Perl6::Pugs>

=head1 AUTHOR

Brian Ingerson C<E<lt>INGY@cpan.orgE<gt>>,
Audrey Tang C<E<lt>autrijus@autrijus.orgE<gt>>

=head1 COPYRIGHT

Copyright (c) 2005. Brian Ingerson. All rights reserved.

This program is free software; you can redistribute it and/or modify it
under the same terms as Perl itself.

See L<http://www.perl.com/perl/misc/Artistic.html>.

=cut

