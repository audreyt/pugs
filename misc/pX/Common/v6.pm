package v6;
use Filter::Simple sub {
    my($this_class,$first_arg,@rest)=@_;
    my ($caller, $file, $line) = caller(1);
    $line++;
    my $location = qq{\n\#line $line "$file"\n};

    my $pkg = ($ENV{PERL5_V6_IMPLEMENTATION}
	       || (defined(%v6::implementation_version)
		   && $v6::implementation_version{$_[0]})
	       || do{ no warnings; $v6::implementation }
	       || 'Pugs::Compiler::Perl6::v6');

    my $cmd = ("use ".$pkg.$first_arg
	       .(@rest ? ",qw(".join(" ",@rest).")" : '')
	       .";\n");
    $_ = $cmd.$location.$_;
};
1;
__END__
