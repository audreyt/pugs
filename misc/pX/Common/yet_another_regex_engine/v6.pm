package v6;
use Filter::Simple sub {
    my($this_class,$first_arg,@rest)=@_;
    my ($caller, $file, $line) = caller(1);
    $line++;
    my $location = qq{\n\#line $line "$file"\n};

    my $pkg = (do{ no warnings; $v6::implementation }
	       || $ENV{PERL5_V6_IMPLEMENTATION}
	       || 'Pugs::Compiler::Perl6::v6');

    my $cmd = ("use ".$pkg.$first_arg
	       .(@rest ? ",qw(".join(" ",@rest).")" : '')
	       .";");
    s/^(use v6-alpha;)/\# $1/
;#	or die "Couldn't find the leading 'use v6-alpha;' to delete it.";
    $_ = $cmd.$location.$_;
#    print STDERR $_,"\n"; $_;
};
1;
__END__
