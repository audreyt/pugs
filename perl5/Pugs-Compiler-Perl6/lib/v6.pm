package v6;
$v6::VERSION = '0.031';

# Documentation in the __END__
use 5.006;
use strict;
use warnings;
use Module::CompileV6-base;
use File::Basename;
# use Pugs::Runtime::Perl6;

binmode(STDOUT, ":utf8");
binmode(STDERR, ":utf8");
binmode(STDIN,  ":utf8");

my $backend_identifier = 'perl5';
my $grammar_identifier;

my $bin;
BEGIN { $bin = ((dirname(__FILE__) || '.') . "/..") };

INIT {
  if($ENV{V6_RESOURCE_GUARD}) {
    require BSD::Resource;
    import BSD::Resource;
    setrlimit(RLIMIT_CPU(), 30, 60) or die "Couldn't setrlimit: $!\n";
    setrlimit(RLIMIT_RSS(), 1048000, 1196000) or die "Couldn't setrlimit: $!\n";
  }
}

sub banner { 
q(         ___ 
        /  __\
 __    _\ \___      _____   _ ___ ___
/\ \  / /\  __ \   /\  __ \/\  __  __ \
\ \ \/ /\ \ \_\ \__\ \ \_\ \ \ \/\ \/\ \
 \ \__/  \ \____/\_).q(\\).q(\ \  __/\ \_\ \_\ \_\  Perl6-in-Perl5
  \/_/    \ ___/\/_/ \ \ \/  \/_/\/_/\/_/  
                      \ \_\              Version: ) . $v6::VERSION . q(
                       \/_/   Copyright 2006, The Pugs Contributors
--------------------------------------------------------------------
 Web: http://pugscode.org/           Email: perl6-compiler@perl.org

);
}

sub pmc_can_output { 1 }

sub pmc_parse_blocks {
    my $class = shift;
    my $text  = shift;
    return [$text, {$class => { use => 'dummy' }}, [$class]]
}

sub pmc_filter {
    my ($class, $module, $line_number, $post_process) = @_;
    return 
        if $module eq '-e';
    $class->SUPER::pmc_filter($module, 0, $post_process);
}

sub pmc_compile {
    my ($class, $source) = @_;

    require Pugs::Compiler::Perl6;

    my $p6 = Pugs::Compiler::Perl6->compile( $source, { backend => $backend_identifier, grammar => $grammar_identifier } );
    my $perl5 = $p6->{perl5};

    # Don't write when we failed to compile, otherwise it never recompiles!
    die unless defined $perl5 
            && length  $perl5;

    # $perl5 =~ s/do\{(.*)\}/$1/s;
    my ($package, $file) = caller(4);

    # allow 'regex' declarations:
    #     use base 'Pugs::Grammar::Base';  
    #   see: t/rules/from_perl6_rules/capture.t

    # postprocess perl5 backend
    if (  $backend_identifier eq 'perl5'
       || $backend_identifier eq 'perl5:Pugs::Emitter::Perl6::Perl5'
       ) {

    $perl5 = 
        ( $package 
            ? "package $package;\n" 
            : "package Main; # no package name\n" ).
        ((!$package or ($package eq 'main')) ? (
            "use Config;\n".
            "use lib split(/\\Q\$Config{path_sep}/, \$ENV{PERL6LIB} || '');\n"
        ) : '').
        "use Scalar::Util;
         use Quantum::Superpositions;
         use Math::Complex;
         use Pugs::Runtime::Perl6;
         use Pugs::Runtime::Perl6Prelude;
         use Pugs::Runtime::Perl5Container;
         use base 'Pugs::Grammar::Base';  
         use strict;
         no strict 'refs';
         no warnings ('void', 'uninitialized');
         \$::_V6_COMPILER_OS      = '$^O';
         \$::_V6_COMPILER_NAME    = 'v6.pm';
         \$::_V6_COMPILER_VERSION = '$v6::VERSION';
         undef \$::_V6_MATCH_;
         my \%_V6_PAD;
         our \%_V6_STATE;
         bool->import();  # True, False
        " .  
        # "Pugs::Runtime::Perl6Prelude->import();\n" .   # XXX - is import() needed?
        $perl5 . "\n" .
        "; 1;\n";

    if ( $ENV{V6TIDY} )
    {
      # Perl::Tidy is used if available
      local $@;   # don't care if there are errors here
      local @ARGV = ();  # "You may not specify any filenames ... - Perl::Tidy.pm
      eval {
        require Perl::Tidy;
        my $perl5_tidy;
        Perl::Tidy::perltidy( 
            source => \$perl5, 
            destination => \$perl5_tidy,
            argv => [
                '--maximum-line-length' => 0,
                '--indent-columns'      => 2,
            ],
        );
        $perl5 = $perl5_tidy;
      }
    }

    }  # / postprocess perl5 backend

    return $perl5;
}

# print "ARGS: \$0='$0' \@ARGV=[ @ARGV ]\n";

if ( @ARGV && !caller ) {
    # We are the main program here
    my ($compile_only, $code);

    if ( @ARGV && $ARGV[0] eq '--compile-only') {
        shift(@ARGV);
        $compile_only++;
    }

    shift(@ARGV) if $ARGV[0] =~ /^--pugs/;
    if ( @ARGV && $ARGV[0] =~ /^-B(.*)/i ) {
        shift(@ARGV);
        $backend_identifier = $1 if $1;
    }
    splice(@ARGV, 0, 2) if @ARGV && $ARGV[0] =~ /^-B$/;

    if ( @ARGV && $ARGV[0] =~ /^-G(.*)/i ) {
        shift(@ARGV);
        $grammar_identifier = $1 if $1;
    }

    while (@ARGV && $ARGV[0] =~ /^-(\w)(.+)/) {
        use Config;
        if($1 eq 'I') {
            $ENV{PERL6LIB} = ((defined($ENV{PERL6LIB}) && $ENV{PERL6LIB} ne '')
                              ? "$Config{path_sep}$ENV{PERL6LIB}"
                              : "");
            $ENV{PERL6LIB} = "$2$ENV{PERL6LIB}";
        }
        shift @ARGV;
    }

    if (@ARGV && $ARGV[0] eq '-v' ) {
        print banner();
        exit 0;
    }

    if (@ARGV && $ARGV[0] =~ s/^-e//) {
        $code = (length($ARGV[0]) ? $ARGV[0] : $ARGV[1]);
    }
    else {
        local $/;
        $code = <>;
    }

    if ($compile_only) {
        print __PACKAGE__->pmc_compile($code);
    }
    else {
        local $@;
        eval __PACKAGE__->pmc_compile($code);
        die $@ if $@;
        exit 0;
    }
}
elsif ( $0 eq '-e' && @ARGV && $ARGV[-1] eq '-v' ) {
    print banner();
    exit 0;
}
elsif ( $0 eq '-e' ) {
    # perl -Ilib -e 'use v6-alpha' ' "hello world".say '
    # perl -Ilib -e 'use v6-alpha' - ' "hello world".say '
    # perl -Ilib -e 'use v6-alpha' - --compile-only ' "hello world".say '
    # perl -Ilib -e 'use v6-alpha' - -Ilib6 ' "hello world".say '
    # echo 42.say | perl -Ilib -e 'use v6-alpha' 
    
    #print "Compile [$0] [@ARGV]\n";
    my ($compile_only, $code);

    shift(@ARGV) if $ARGV[0] && $ARGV[0] eq '-';

    if ($ARGV[0] && $ARGV[0] eq '--compile-only') {
        shift(@ARGV);
        $compile_only++;
    }

    shift(@ARGV) if $ARGV[0] && $ARGV[0] =~ /^--pugs/;
    if ( @ARGV && $ARGV[0] =~ /^-B(.*)/i ) {
        shift(@ARGV);
        $backend_identifier = $1 if $1;
    }
    splice(@ARGV, 0, 2) if $ARGV[0] && $ARGV[0] =~ /^-B$/;

    if ( @ARGV && $ARGV[0] =~ /^-G(.*)/i ) {
        shift(@ARGV);
        $grammar_identifier = $1 if $1;
    }

    while (@ARGV && $ARGV[0] =~ /^-(\w)(.+)/) {
        use Config;
        if($1 eq 'I') {
            $ENV{PERL6LIB} = ((defined($ENV{PERL6LIB}) && $ENV{PERL6LIB} ne '')
                              ? "$Config{path_sep}$ENV{PERL6LIB}"
                              : "");
            $ENV{PERL6LIB} = "$2$ENV{PERL6LIB}";
        }
        shift @ARGV;
    }

    if (@ARGV) {  # and $ARGV[0] =~ s/^-e//) {
        $code = (length($ARGV[0]) ? $ARGV[0] : $ARGV[1]);
    }
    else {
        local $/;
        $code = <>;
    }

    if ($compile_only) {
        print __PACKAGE__->pmc_compile($code);
    }
    else {
        local $@;
        eval __PACKAGE__->pmc_compile($code);
        die $@ if $@;
        exit 0;
    }
}

1;

__END__

=head1 NAME 

v6 - An experimental Perl 6 implementation

=head1 SYNOPSIS

    # file: hello_world.pl
    use v6-alpha;
    "hello, world".say;

    $ perl hello_world.pl

C<v6-alpha> can be used in one-liners:

    $ perl -e 'use v6-alpha' ' 42.say '
    $ perl -e 'use v6-alpha' - ' 42.say '
    $ perl -e 'use v6-alpha' - --compile-only ' 42.say '
    $ perl -e 'use v6-alpha' - -Ilib6 ' 42.say '
    $ echo 42.say | perl -e 'use v6-alpha' 

C<v6.pm> can also be used as a plain program. 
This examples assume that v6.pm is in the C<./lib> directory:

    $ perl lib/v6.pm -e 'for 1,2,3 -> $x { say $x }'
    $ perl lib/v6.pm --compile-only -e '<hello>.say;'

Print the version banner:

    $ perl -e 'use v6-alpha' - -v
    $ perl lib/v6.pm -v

=head1 DESCRIPTION

The C<v6> module is a front-end to the experimental Perl6-to-Perl5 compiler.

The current state of this compiler implementation only provides a small sample of
Perl 6 syntax and semantics.

=head2 Other Perl 6 implementations

C<Pugs> Perl 6 is currently the most complete implementation. Pugs is written on top of Haskell.

Perl 6 on C<Parrot> aims to become the best performing implementation.
Parrot is a virtual machine designed to efficiently compile and execute bytecode for interpreted languages.

C<v6.pm> is completely independent of Pugs or Parrot. 

=head2 Compile-time system

L<Pugs::Compiler::Rule> provides an implementation for Perl 6 Rules, which are used to define the Perl 6 grammar.

L<Parse::Yapp> is used for implementing the operator precedence parser.

L<Module::Compile> and L<Cache::Cache> provide the precompilation cache
infrastructure.

=head2 Runtime system

The object system is provided by L<Moose>. 
Moose is an extension of the Perl 5 object system.
Moose is built on top of L<Class::MOP>, which is a metaclass system for Perl 5.

L<Data::Bind> and L<Sub::Multi> implement the semantics for perl6-style variable binding, as well as subroutine call argument passing and binding, in Perl 5.

=head1 REQUIREMENTS

- The source file header must be valid perl5 I<and> perl6 code.

This is a valid header:

    #!/usr/bin/perl
    use v6-alpha;

* it executes perl5

* perl5 will call the C<v6.pm> module.

This is an invalid header:

    #!/usr/bin/pugs
    use v6;

* it tells perl5 to execute C</usr/bin/pugs>.

* it tells perl5 that Perl v6.0.0 required.

- The C<Pugs::Compiler::Rule> module must be properly installed.

An improperly installed C<Pugs::Compiler::Rule> module would prevent the Perl 6 compiler
from bootstrapping. 

If that is the case, running C<Makefile.PL> and C<make> in
C<Pugs::Compiler::Rule> should fix the problem.

- The perl5 executable must have PMC support. 

PMC support is required for loading precompiled Perl 6 files.

If you see the error below, it may happen that your perl was compiled without PMC support.

  Can't locate object method "compile" via package "Pugs::Compiler::Perl6"

Please see L<http://rt.cpan.org/Public/Bug/Display.html?id=20152>

=head1 ENVIRONMENT VARIABLES

* PERL6LIB

Same usage as PERL5LIB - sets the search path for Perl 6 modules.

* V6DUMPAST

If set, the compiler will dump the syntax tree to C<STDOUT> just before emitting code, using C<Data::Dumper>.

* V6TIDY

If set, the compiler output will be much more readable, but there will
be a lot of slowdown in compiler speed.

=head1 COMMAND LINE SWITCHES

* --compile-only

When using C<v6.pm> from the command line, dumps the emitted code
to C<STDOUT> and then exit:

    $ perl -e 'use v6-alpha' - --compile-only ' 42.say '

    $ perl -Ilib lib/v6.pm --compile-only -e '<hello>.say;'

* -B

Selects alternate code generation backends.

The default is '-Bperl5:Pugs::Emitter::Perl6::Perl5'.
'-Bperl5' also invokes the default backend.

    $ perl -e 'use v6-alpha' - --compile-only -Bperl5:MyEmitter ' 42.say '

    $ perl -e 'use v6-alpha' - --compile-only -BPugs::Emitter::YAML ' 123 '
    --- 
    statements: 
      - 
        int: 123
        pos: 1

The backend module must provide the C<emit($grammar, $ast)> subroutine.

* -G

Selects alternate grammar frontends.

The default is '-Gperl5:Pugs::Grammar::Perl6'.

=head1 AUTHORS

The Pugs Team E<lt>perl6-compiler@perl.orgE<gt>.

=head1 SEE ALSO

The Perl 6 homepage at L<http://dev.perl.org/perl6>.

- the Perl 6 Synopsis: L<http://dev.perl.org/perl6/doc/synopsis.html>.

The Pugs homepage at L<http://pugscode.org/>.

The Parrot homepage at L<http://www.parrotcode.org>.

=head1 COPYRIGHT

Copyright 2006 by Flavio Soibelmann Glock and others.

This program is free software; you can redistribute it and/or modify it
under the same terms as Perl itself.

See L<http://www.perl.com/perl/misc/Artistic.html>

=cut
