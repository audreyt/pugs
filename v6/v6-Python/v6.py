#!/usr/bin/env python

import MiniPerl.Parse as parse
import sys

if len(sys.argv) > 1:
	data = open(sys.argv[1]).read()
else:
	data = sys.stdin.read()

print MiniPerl.Parse.yacc.parse(data)
