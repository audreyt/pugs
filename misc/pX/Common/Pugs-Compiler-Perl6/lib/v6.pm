package v6;
$v6::VERSION = '0.01_001';

# Documentation in the __END__
use 5.006;
use strict;
use warnings;
use Module::Compile-base;
use File::Basename;
use Pugs::Runtime::Perl6;

my $bin;
BEGIN { $bin = ((dirname(__FILE__) || '.') . "/..") };
use lib (
    "$bin/lib",
    "$bin/../Pugs-Compiler-Rule/lib",
);

sub pmc_can_output { 1 }

sub pmc_compile {
    my ($class, $source) = @_;

    my $file = (caller(4))[1];
    if (defined $file and $file !~ /\.pm$/i) {
        # Do the freshness check ourselves
        my $pmc = $file.'c';
        my $pmc_is_uptodate = (-s $pmc and (-M $pmc <= -M $file));
        if ($pmc_is_uptodate) {
            local $@; do $pmc; die $@ if $@; exit 0;
        }
    }

    require Pugs::Compiler::Perl6;

    my $p6 = Pugs::Compiler::Perl6->compile( $source );
    my $perl5 = $p6->{perl5};

    # $perl5 =~ s/do\{(.*)\}/$1/s;
    $perl5 = 
        "use Pugs::Runtime::Perl6;\n" . 
        "use strict;\n" . 
        #"no warnings 'array';\n" . 
        $perl5 . "\n";

    {
      # Perl::Tidy is used if available
      local $@;   # don't care if there are errors here
      local @ARGV = ();  # "You may not specify any filenames ... - Perl::Tidy.pm
      eval {
        require Perl::Tidy;
        my $perl5_tidy;
        Perl::Tidy::perltidy( source => \$perl5, destination => \$perl5_tidy );
        $perl5 = $perl5_tidy;
      }
    }

    return $perl5;
}

if (@ARGV and !caller) {
    # We are the main program here
    my ($compile_only, $code);

    if ($ARGV[0] eq '--compile-only') {
        shift(@ARGV);
        $compile_only++;
    }

    shift(@ARGV) if $ARGV[0] =~ /^--pugs/;
    shift(@ARGV) if $ARGV[0] =~ /^-Bperl5$/i;
    splice(@ARGV, 0, 2) if $ARGV[0] =~ /^-B$/;

    if (@ARGV and $ARGV[0] =~ s/^-e//) {
        $code = (length($ARGV[0]) ? $ARGV[0] : $ARGV[1]);
    }
    else {
        local $/;
        $code = <>;
    }

    if ($compile_only) {
        print __PACKAGE__->pmc_compile($code);
    }
    else {
        local $@;
        eval __PACKAGE__->pmc_compile($code);
        die $@ if $@;
        exit 0;
    }
}

1;

__END__

=head1 NAME 

v6 - an experimental Perl 6 implementation (under way)

=head1 SYNOPSIS

Command line:

    $ perl -Ilib lib/v6.pm -e ' for 1,2,3 -> $x { say $x }'

Compile-only:

    $ perl -Ilib lib/v6.pm --compile-only -e ' <hello>.say; '

Script or module:

    # file: hello_world.pl
    use v6-pugs;
    "hello, world".say;

    $ perl hello_world.pl

=head1 DESCRIPTION

The C<v6> module is a front-end to the experimental Perl6-to-Perl5 compiler.

The current state of this compiler implementation only provide a small sample of
Perl 6 syntax and semantics.

=head2 Other Perl 6 implementations

The Pugs/Haskell Perl 6 is currently the most complete implementation. 
Pugs currently has some issues with Perl 5 interoperability.

Parrot Perl 6 is the best performing implementation by far.
The Parrot implementation is currently at a comparable state as v6.pm.

Although running C<v6> requires the installation of a lot of Perl 5 modules,
it is completely independent of Pugs or Parrot. 

=head1 AUTHORS

The Pugs Team E<lt>perl6-compiler@perl.orgE<gt>.

=head1 SEE ALSO

The Perl 6 Synopsis: L<http://dev.perl.org/perl6/doc/synopsis.html>

=head1 COPYRIGHT

Copyright 2006 by Flavio Soibelmann Glock and others.

This program is free software; you can redistribute it and/or modify it
under the same terms as Perl itself.

See L<http://www.perl.com/perl/misc/Artistic.html>

=cut
