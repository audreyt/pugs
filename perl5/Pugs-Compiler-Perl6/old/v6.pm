package v6;

# invokes the Perl6-to-Perl5 compiler and creates a .pmc file

use strict;
use warnings;

my $filename = (caller)[1];
my $pmc = $filename;
$pmc =~ s/\.(pl|pm|t)$/\.pmc/;
die "this is not a .pl, .t or .pm file"
    unless $pmc =~ /\.pmc/;
# test file dates
my $pmc_is_uptodate = (-s $pmc and (-M $pmc <= -M $filename));
# print "up to date: $pmc_is_uptodate\n";
if ( $pmc_is_uptodate ) {
    do $pmc 
        unless $filename =~ /\.pm$/;
    exit 0;
    # return 1;
}

my $p6;
# delay use'ing until we need it
eval 
  q(
    use FindBin '$Bin';
    use lib
        "$Bin/lib",
        "$Bin/../Pugs-Compiler-Rule/lib",
        "$Bin/../Pugs-Utils/lib",
        #"$Bin/../Pugs-Compiler-Precedence/lib",
        
        "$Bin/../lib",
        "$Bin/../../Pugs-Compiler-Rule/lib",
        "$Bin/../../Pugs-Utils/lib",
        #"$Bin/../../Pugs-Compiler-Precedence/lib",
    ;

    use Pugs::Compiler::Perl6;
    use Data::Dumper;

    # print "P6 loaded from '$filename'\n";
    open FILE, "<", $filename
        or die "can't open $filename for compilation";
    my @a=<FILE>;
    close FILE;
    my $src = join('', @a);
    $p6 = Pugs::Compiler::Perl6->compile( $src );
    #use YAML;
    #print Dump $match->();
    #print dump_tree $p6;
    #print Dumper $p6;
    # print "tail: ", substr( ${$match}->{tail}, 0, 20 ),"...\n";
    #print "P5: ",$p6->{perl5}, "\n";

    open FILE, ">", $pmc
        or die "can't open $pmc for writing";

    print FILE $p6->{perl5}, "\n";

    # print "created .pmc file: $^X $pmc \n";
  );
warn "compilation error: ", $@ if $@;
eval $p6->{perl5}  
    unless $filename =~ /\.pm$/;
exit 0;

1;
