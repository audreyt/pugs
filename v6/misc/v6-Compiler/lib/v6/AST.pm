
use v6-alpha;

module v6::AST;

# Note: v6.pm uses Pugs::Runtime::Perl6AST instead

sub node ( $match, $node_type ) {
    $match does $node_type
}
