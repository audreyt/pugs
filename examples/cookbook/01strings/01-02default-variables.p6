#!/usr/bin/perl6

use v6;

=pod

=head1 Default values

You want to provide default values for false or undefined variables.

    my ($true,$false,$string,$defined,$last);

    # Provide a default for undefined values

    $false //= 0; 

    # The default for $string will be 'string', 
    # unless the left-hand side of || evaluates 
    # as True.
    # Similar to Perl 5, use with caution. 

    $string = $false || 'string'; # 0 is (normally) Bool::False: 
                                  # thus default is 'string' 
    
    # Default to the first defined value in a list
    # use the reduce operator

    $defined = [//] ($true,$false,$string,$defined,$last); # 0 
    
    # Default to the first true value in a list 
    # use the reduce operator

    $true = [||] ($true,$false,$string,$defined,$last); # 'string' 

    # To provide the last defined value in a list, as default,
    # define your own right-hand side definedness test
    # use the reduce operator

    sub infix:<\\>($lhs,$rhs){ $rhs // $lhs };
    $last = [\\] ($true,$false,$string,$defined,$last); # 0  
    
    # force strict boolean context on a variable 

    my $y is 0 but true;
    $y ||= 4;
    say $y; # 0

    my $z is 'Ovid' but false;
    $z ||= 4;
    say $z; # 4
    
=cut

my ($true,$false,$string,$defined,$last);

$false //= 0; 
say "Default for undefined: $false"; 

$string = $false || 'string'; # 0 is (normally) False: 
say "Default if False: $string"; # string 

$defined = [//] ($true,$false,$string,$defined,$last); # 0 
say "Default is first defined value of list: $defined";

$true = [||] ($true,$false,$string,$defined,$last); # 'string' 
say "Default is first True value of list: $true";

sub infix:<\\>($lhs,$rhs){ $rhs // $lhs ;}
$last = [\\] ($true,$false,$string,$defined,$last); # 0  
say "Default is last defined value of List: $last";

#my $y is 0 but true;
# TODO Error

#my $z is 'Ovid' but false;
# TODO Error

