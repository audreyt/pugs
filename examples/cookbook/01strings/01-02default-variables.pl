use v6-alpha;

=pod

=head1 Default values

You want to provide default values for false or undefined variables.

    my ($true,$false,$string,$defined,$last);

    # Provide a default for undefined variables

    $false //= 0; 

    # The default for $string will be 'string', 
    # unless the left-hand side of || evaluates 
    # as True.
    # Similar to Perl 5, use with caution. 

    $string ||= $false || 'string'; # 0 is (normally) Bool::False: 
                                  # thus default is 'string' 
    
    # Default to the first defined value in a list
    # use the reduce operator

    $defined //= [//] ($true,$false,$string,$defined,$last); # 0 
    
    # Default to the first true value in a list 
    # use the reduce operator

    $true ||= [||] ($true,$false,$string,$defined,$last); # 'string' 

    # To provide the last defined value in a list, as default,
    # define your own right-hand side definedness test
    # use the reduce operator

    sub infix:<\\>($lhs,$rhs){ $rhs // $lhs };
    $last //= [\\] ($true,$false,$string,$defined,$last); # 0  

    # Or ... 
    $last=undef;

    # Reverse the array and use the builtin definedness test
    # use the reduce operator
    $last //= [//] reverse($true,$false,$string,$defined,$last); # 0  
    
    # force strict boolean context on a variable 

    my $y is 0 but true;
    $y ||= 4;
    say $y; # 0

    my $z is 'Ovid' but false;
    $z ||= 4;
    say $z; # 4
    
=cut

my ($true,$false,$string,$defined,$last);

# Provide a default for undefined variables

$false //= 0; 
say "Default for undefined: $false"; 

$string ||= $false || 'string'; 
say "Default if false: $string"; # string 

$defined //= [//] ($true,$false,$string,$defined,$last); # 0 
say "Default to first defined value of a list: $defined";

$true ||= [||] ($true,$false,$string,$defined,$last); # 'string' 
say "Default to first True value of a list: $true";

sub infix:{qn'\\'} ($lhs,$rhs){ $rhs // $lhs };
$last //= [\\] ($true,$false,$string,$defined,$last); # 0  
say "Default to last defined value of a list: $last";

# Or ... 
$last=undef;
$last //= [//] reverse($true,$false,$string,$defined,$last); # 0  
say "Default to last defined value of a list: $last";

#my $y is 0 but true;
# TODO Error

#my $z is 'Ovid' but false;
# TODO Error

