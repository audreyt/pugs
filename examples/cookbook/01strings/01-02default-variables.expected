Default for undefined: 0
Default if false: string
Default to first defined value of a list: 0
Default to first True value of a list: string
Default to last defined value of a list: 0
Default to last defined value of a list: 0
